ingestion-edge/ingestion_edge/publish.py

```diff
@@ -3,44 +3,135 @@
+from google.api_core.exceptions import Forbidden
+from google.api_core.exceptions import RetryError
+from google.api_core.exceptions import ServerError
+from google.api_core.exceptions import TooManyRequests
+from google.cloud.pubsub_v1 import PublisherClient
+from google.cloud.pubsub_v1.publisher.exceptions import PublishError
+from google.cloud.pubsub_v1.types import PubsubMessage
+from persistqueue import SQLiteAckQueue
+from persistqueue.exceptions import Empty
 from typing import Tuple
+import json
+
+FLUSH_EMPTY_STATUS = 204
+FLUSH_COMPLETE_STATUS = 200
+FLUSH_INCOMPLETE_STATUS = 504
+# Errors where message should be queued locally after pubsub delivery fails
+# To minimize data loss include errors that may require manual intervention
+TRANSIENT_ERRORS = (
+    # The API thinks the client should retry
+    Aborted,
+    # The API was interrupted when handling the request
+    Cancelled,
+    # The PublisherClient lacks permissions and may require manual intervention
+    Forbidden,
+    # API call ran out of retries
+    RetryError,
+    # API outage or connection issue
+    ServerError,
+    # A python Future timed out
+    TimeoutError,
+    # The API is throttling requests
+    TooManyRequests,
+)
 
 publish = Blueprint('publish', __name__)
 client = PublisherClient()
+q = SQLiteAckQueue(QUEUE_PATH)
+
+
+def _publish(topic, data, attrs) -> str:
+    """Publish one message to the topic
+
+    Uses PublisherClient to try to publish the message in a batch and handles
+    non-transient batch failure by retrying synchronously with a single message
+    """
+    try:
+        return client.publish(topic, data, **attrs).result()
+    except PublishError:
+        # Batch failed because pubsub rejected at least one message
+        # retry this message alone to determine if it was rejected
+        response = client.api.publish(
+            topic=topic,
+            messages=[PubsubMessage(data=data, attributes=attrs)],
+        )
+        if len(response.message_ids) != 1:
+            raise PublishError("message was not successfully published")
+        return response.message_ids[0]
+
+
+@publish.route("/__flush__")
+def flush_queue() -> Tuple[str, int]:
+    """Flush messages from the local queue to pubsub
+
+    Call periodically on each docker container to ensure reliable delivery
+
+    Access to this endpoint should be restricted
+    """
+    if q.size == 0:
+        # early return for empty queue
+        return "", FLUSH_EMPTY_STATUS
+    done = 0
+    # while queue has messages ready to process
+    while q.size > 0:
+        try:
+            message = q.get(block=False)
+        except Empty:
+            # queue is empty or sqlite needs a retry
+            continue
+        try:
+            try:
+                _publish(*message)
+            except:  # noqa: E722
+                # nack then re-raise on absolutely any exception
+                q.nack(message)
+                raise
```

Comment:
I think this works as written, but the nested `try` is a bit difficult to parse. I think I'd prefer duplication of the `q.nack` call so that this can be flatter:

```
try:
  _publish(*message)
except TRANSIENT_ERRORS:
  q.nack(message)
  response = ...
  return ...
except:
  q.nack(message)
  raise
else:
  q.ack(message)
  done += 1
```